### ISSUE TITLE
Often and Inconsistent CODEC_UPLINK Error

### ISSUE THREAD

**Issue by teterski** [chirpstack/chirpstack#264]
# Often and Inconsistent CODEC_UPLINK Error

Hello, we are setting up Chirpstack v4 and are running into a few issues not encountered in v3.

For troubleshooting purposes, I have a device set up with a very basic codec, the code for which is below:
``` js
//This is a test codec to debug codec related issues in Chirpstack v4.
function decodeUplink(input) {
    return {
        data: {
            test: 'TEST'
        }
    };
}

function encodeDownlink(input) {
    return {
        bytes: []
    };
}
```

The issue is that on an inconsistent basis, but quiet often (sometimes almost all the time), we are getting `UPLINK_CODEC` errors. However, at times, the event goes through without any issues, and we receive our desired `Object` in the event with the `test:'TEST'` field in it.  When the issue occurs, the events are received in pairs, one `up` event and one `log` event with the error in it. Here is an example of these events

Log event:
```json
{
    "time": "2023-08-23T17:04:52.691922687+00:00",
    "deviceInfo": {
        "tenantId": "52f14cd4-c6f1-4fbd-8f87-4025e1d49242",
        "tenantName": "ChirpStack",
        "applicationId": "ca2dedf4-8a6e-419a-a84e-72b6c061c851",
        "applicationName": "Test PLS",
        "deviceProfileId": "92c1afef-b784-41cc-b907-f6282e7a8356",
        "deviceProfileName": "ABP_PLS-Canonical",
        "deviceName": "Test Guppy",
        "devEui": "70b3d57050007221",
        "deviceClassEnabled": "CLASS_A",
        "tags": {}
    },
    "level": "ERROR",
    "code": "UPLINK_CODEC",
    "description": "Exception generated by quickjs",
    "context": {
        "deduplication_id": "31ee17a5-9010-4ccf-a59b-b419aa8b310b"
    }
}
```
As you can see the description does not provide the actual exception generated by quickjs.

Up event:
```json
{
    "deduplicationId": "31ee17a5-9010-4ccf-a59b-b419aa8b310b",
    "time": "2023-08-23T17:04:52.269820317+00:00",
    "deviceInfo": {
        "tenantId": "52f14cd4-c6f1-4fbd-8f87-4025e1d49242",
        "tenantName": "ChirpStack",
        "applicationId": "ca2dedf4-8a6e-419a-a84e-72b6c061c851",
        "applicationName": "Test PLS",
        "deviceProfileId": "92c1afef-b784-41cc-b907-f6282e7a8356",
        "deviceProfileName": "ABP_PLS-Canonical",
        "deviceName": "Test Guppy",
        "devEui": "70b3d57050007221",
        "deviceClassEnabled": "CLASS_A",
        "tags": {}
    },
    "devAddr": "50007221",
    "adr": false,
    "dr": 0,
    "fCnt": 13654,
    "fPort": 1,
    "confirmed": false,
    "data": "14Y=",
    "rxInfo": [
        {
            "gatewayId": "0016c001f1529cb0",
            "uplinkId": 28229,
            "rssi": -48,
            "snr": 13.5,
            "channel": 5,
            "rfChain": 1,
            "location": {},
            "context": "XC7t2w==",
            "metadata": {
                "region_common_name": "US915",
                "region_config_id": "us915_1"
            },
            "crcStatus": "CRC_OK"
        }
    ],
    "txInfo": {
        "frequency": 904900000,
        "modulation": {
            "lora": {
                "bandwidth": 125000,
                "spreadingFactor": 10,
                "codeRate": "CR_4_5"
            }
        }
    }
}
```

When running `journalctl -f -u chirpstack` we get the following logs for the period of time during which the event occurred. I am not seeing anything that would hint at what could be causing this issue.
```log
Aug 23 11:04:49 SED1 chirpstack[358]: 2023-08-23T17:04:49.178560Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_config_id="us915_1" topic="us915_1/gateway/0016c001f1529cb0/event/up" qos=0 json=false
Aug 23 11:04:49 SED1 chirpstack[358]: 2023-08-23T17:04:49.386358Z  INFO up{deduplication_id=28183715-6ea8-460d-a21d-9b0e282d880d}: chirpstack::uplink: Uplink received m_type="UnconfirmedDataUp"
Aug 23 11:04:49 SED1 chirpstack[358]: 2023-08-23T17:04:49.402901Z  WARN up{deduplication_id=28183715-6ea8-460d-a21d-9b0e282d880d}:data_up: chirpstack::uplink::data: No device-session exists for dev_addr dev_addr=00000022
Aug 23 11:04:51 SED1 chirpstack[358]: 2023-08-23T17:04:51.477663Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_config_id="us915_1" topic="us915_1/gateway/0016c001f1529cb0/event/up" qos=0 json=false
Aug 23 11:04:51 SED1 chirpstack[358]: 2023-08-23T17:04:51.684540Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}: chirpstack::uplink: Uplink received m_type="UnconfirmedDataUp"
Aug 23 11:04:51 SED1 chirpstack[358]: 2023-08-23T17:04:51.698787Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::device_session: Device-session saved dev_eui=70b3d57050007221 dev_addr=50007221
Aug 23 11:04:51 SED1 chirpstack[358]: 2023-08-23T17:04:51.733529Z  INFO chirpstack::storage::device: Setting device lock dev_eui=70b3d57050007221
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.253447Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::device: Data-rate updated dev_eui=70b3d57050007221 dr=0
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.269217Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::device_gateway: Gateway rx-info saved dev_eui=70b3d57050007221
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.733141Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::device_session: Device-session saved dev_eui=70b3d57050007221 dev_addr=50007221
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.745175Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::metrics: Metrics saved name=device:70b3d57050007221 aggregation=HOUR
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.748749Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::metrics: Metrics saved name=device:70b3d57050007221 aggregation=DAY
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.752598Z  INFO up{deduplication_id=31ee17a5-9010-4ccf-a59b-b419aa8b310b}:data_up: chirpstack::storage::metrics: Metrics saved name=device:70b3d57050007221 aggregation=MONTH
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.768587Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_config_id="us915_1" topic="us915_1/gateway/0016c001f1529cb0/event/up" qos=0 json=false
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.812230Z  INFO chirpstack::integration::mqtt: Publishing event topic=application/ca2dedf4-8a6e-419a-a84e-72b6c061c851/device/70b3d57050007221/event/up
Aug 23 11:04:52 SED1 chirpstack[358]: 2023-08-23T17:04:52.822867Z  INFO chirpstack::integration::mqtt: Publishing event topic=application/ca2dedf4-8a6e-419a-a84e-72b6c061c851/device/70b3d57050007221/event/log
Aug 23 11:04:53 SED1 chirpstack[358]: 2023-08-23T17:04:53.005092Z  INFO up{deduplication_id=aae8e651-527c-4c96-81a8-8efe9feace73}: chirpstack::uplink: Uplink received m_type="UnconfirmedDataUp"
Aug 23 11:04:53 SED1 chirpstack[358]: 2023-08-23T17:04:53.005559Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_config_id="us915_1" topic="us915_1/gateway/0016c001f1529cb0/event/up" qos=0 json=false
Aug 23 11:04:53 SED1 chirpstack[358]: 2023-08-23T17:04:53.033404Z  WARN up{deduplication_id=aae8e651-527c-4c96-81a8-8efe9feace73}:data_up: chirpstack::uplink::data: No device-session exists for dev_addr dev_addr=5000900f
Aug 23 11:04:53 SED1 chirpstack[358]: 2023-08-23T17:04:53.224426Z  INFO up{deduplication_id=0304d5bd-195a-4db2-9668-6d722707e57c}: chirpstack::uplink: Uplink received m_type="ConfirmedDataUp"
Aug 23 11:04:53 SED1 chirpstack[358]: 2023-08-23T17:04:53.248528Z  WARN up{deduplication_id=0304d5bd-195a-4db2-9668-6d722707e57c}:data_up: chirpstack::uplink::data: No device-session exists for dev_addr dev_addr=50009212
Aug 23 11:04:54 SED1 chirpstack[358]: 2023-08-23T17:04:54.501662Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_config_id="us915_1" topic="us915_1/gateway/0016c001f1529cb0/event/up" qos=0 json=false
Aug 23 11:04:54 SED1 chirpstack[358]: 2023-08-23T17:04:54.708097Z  INFO up{deduplication_id=f78cf7d6-3dca-46ff-a5a4-27e95110106f}: chirpstack::uplink: Uplink received m_type="UnconfirmedDataUp"
Aug 23 11:04:54 SED1 chirpstack[358]: 2023-08-23T17:04:54.725540Z  WARN up{deduplication_id=f78cf7d6-3dca-46ff-a5a4-27e95110106f}:data_up: chirpstack::uplink::data: No device-session exists for dev_addr dev_addr=01021280
```

The issue occurs on multiple different device types, with different profiles and different codecs. The issue occurs regardless if there is an integration set up for the application or not. Has anyone encountered anything similar, or has any idea on how we could go about troubleshooting this or resolving this issue.

**Comment by teterski** [comment]
Forgot to note that we are using Chirpstack v4.4.3 which is the latest version right now.

**Comment by teterski** [comment]
It appears the errors were caused by performance issues on the device we were running Chirpstack on. We noticed that when frames would come in, our device's CPU usage would reach 100% when trying to decode the frame, and we suspect it resulted in the quickjs engine timing out with the error due to it taking too long to decode. After turning off some unneeded services on the device and fixing the performance issues, we no longer receive the error.
